libraries/map/src/main/kotlin/io/element/android/libraries/map/MapView.kt

```diff
@@ -0,0 +1,263 @@
+import io.element.android.libraries.designsystem.theme.components.Icon
+import io.element.android.libraries.map.internal.buildTileServerUrl
+import kotlinx.collections.immutable.ImmutableList
+import kotlinx.collections.immutable.toImmutableList
+import timber.log.Timber
+import kotlin.coroutines.resume
+import kotlin.coroutines.suspendCoroutine
+
+/**
+ * Composable wrapper around MapLibre's [MapView].
+ */
+@SuppressLint("MissingPermission")
+@Composable
+fun MapView(
+    modifier: Modifier = Modifier,
+    mapState: MapState = rememberMapState(),
+    darkMode: Boolean = !ElementTheme.colors.isLight,
+    onLocationClick: () -> Unit,
+) {
+    // When in preview, early return a Box with the received modifier preserving layout
+    if (LocalInspectionMode.current) {
+        Box(modifier = modifier)
+        return
+    }
+
+    val context = LocalContext.current
+    val mapView = remember {
+        Mapbox.getInstance(context)
+        MapView(context)
+    }
+    var mapRefs by remember { mutableStateOf<MapRefs?>(null) }
+
+    // Build map
+    LaunchedEffect(darkMode) {
+        mapView.awaitMap().let { map ->
+            map.uiSettings.apply {
+                isCompassEnabled = false
+            }
+            map.setStyle(buildTileServerUrl(darkMode = darkMode)) { style ->
+                mapRefs = MapRefs(
+                    map = map,
+                    symbolManager = SymbolManager(mapView, map, style).apply {
+                        iconAllowOverlap = true
+                    },
+                    style = style
+                )
+            }
+        }
+    }
+
+    // Update state position when moving map
+    LaunchedEffect(mapRefs) {
+        mapRefs?.let {
+            // TODO: Are we actually leaking memory by not removing this listener?
+            it.map.addOnCameraMoveListener {
+                it.map.cameraPosition.target?.let { target ->
+                    val position = Position(
+                        lat = target.latitude,
+                        lon = target.longitude,
+                        zoom = it.map.cameraPosition.zoom
+                    )
+                    mapState.position = position
+                    Timber.d("Camera moved to: $position")
+                }
+            }
+        }
+    }
+
+    // Move map to given position when state has changed
+    LaunchedEffect(mapRefs, mapState.position) {
+        mapRefs?.map?.moveCamera(
+            CameraUpdateFactory.newCameraPosition(
+                CameraPosition.Builder()
+                    .target(LatLng(mapState.position.lat, mapState.position.lon))
+                    .zoom(mapState.position.zoom).build()
+            )
+        )
+        Timber.d("Camera position updated to: ${mapState.position}")
+    }
+
+    // Draw pin
+    LaunchedEffect(mapRefs, mapState.location) {
+        // Intentional shadowing
+        @Suppress("NAME_SHADOWING") val mapRefs = mapRefs ?: return@LaunchedEffect
```

Comment:
Would wrapping this code in `mapRefs?.let { ... }` remove the need for this?

Reply:
Wow yes! Thanks, sometimes my brain just goes sideways...